File: lab-starters/lab4/src/ingest.py
"""
ingest.py -- Market data identifier ingestion pipeline

Reads a CSV file of market data records, resolves each instrument
identifier to its OpenFIGI code using FigiClient, counts lookup
frequency by exchange code, and writes enriched records to stdout as CSV.

Usage:
    python src/ingest.py data/sample_input.csv > data/python_output.csv
"""

import csv
import logging
import sys
from collections import Counter
from pathlib import Path
from typing import Optional

from src.figi_client import FigiClient, FigiClientError

logging.basicConfig(level=logging.INFO, format="%(levelname)s %(message)s")
logger = logging.getLogger(__name__)

CRITICAL_FIELDS = {"instrument_id", "exchange_code", "price", "volume", "record_id"}


def load_records(input_path: Path) -> list[dict]:
    """Load market data records from a CSV file.

    Args:
        input_path: Path to the input CSV file.

    Returns:
        List of record dicts with string values for all columns.

    Raises:
        FileNotFoundError: If input_path does not exist.
        ValueError: If the CSV file contains no records.
    """
    logger.info(f"Starting load_records from {input_path}")

    if not input_path.exists():
        raise FileNotFoundError(f"Input file not found: {input_path}")

    records: list[dict] = []
    with input_path.open(newline="", encoding="utf-8") as fh:
        reader = csv.DictReader(fh)
        for row in reader:
            for field in CRITICAL_FIELDS:
                if row.get(field) is None:
                    logger.warning(
                        f"Record {row.get('record_id', '?')}: missing critical field {field}"
                    )
            records.append(dict(row))

    if not records:
        raise ValueError(f"No records found in {input_path}")

    logger.info(f"Completed load_records: {len(records)} records loaded")
    return records


def resolve_figis(
    records: list[dict],
    client: FigiClient,
) -> dict[str, str]:
    """Resolve instrument identifiers to OpenFIGI codes.

    Args:
        records: List of market data record dicts.
        client:  Configured FigiClient instance.

    Returns:
        Mapping of instrument_id to FIGI string (or 'UNKNOWN' on failure).
    """
    logger.info(f"Starting resolve_figis with {len(records)} records")

    identifiers = [
        {
            "idType": r.get("id_type") or "TICKER",
            "idValue": r["instrument_id"],
            "exchCode": r.get("exchange_code") or None,
        }
        for r in records
        if r.get("instrument_id")
    ]

    try:
        results = client.do_request(identifiers)
    except FigiClientError as exc:
        logger.error(f"FigiClient request failed: {exc}")
        results = None

    figi_map: dict[str, str] = {}
    if results:
        for i, item in enumerate(results):
            instrument_id = identifiers[i]["idValue"]
            data = item.get("data") if item else None
            figi_map[instrument_id] = data[0]["figi"] if data else "UNKNOWN"
    else:
        for ident in identifiers:
            figi_map[ident["idValue"]] = "UNKNOWN"

    logger.info(f"Completed resolve_figis: {len(figi_map)} identifiers resolved")
    return figi_map


def rank_exchanges(records: list[dict]) -> tuple[Counter, dict[str, int]]:
    """Count and rank exchanges by lookup frequency.

    Exchanges with equal counts are ranked by exchange_code ascending,
    so the ranking is deterministic regardless of input order.

    Args:
        records: List of market data record dicts.

    Returns:
        Tuple of (exchange_counts Counter, exchange_rank dict).
    """
    logger.info(f"Starting rank_exchanges with {len(records)} records")

    exchange_counts: Counter = Counter(
        r.get("exchange_code") or "UNKNOWN" for r in records
    )

    # Sort descending by count, then by exchange_code so ties are deterministic.
    sorted_exchanges = sorted(exchange_counts, key=lambda e: (-exchange_counts[e], e))

    exchange_rank = {exch: rank + 1 for rank, exch in enumerate(sorted_exchanges)}

    logger.info(f"Completed rank_exchanges: {len(exchange_counts)} exchanges ranked")
    return exchange_counts, exchange_rank


def write_output(
    records: list[dict],
    figi_map: dict[str, str],
    exchange_counts: Counter,
    exchange_rank: dict[str, int],
) -> None:
    """Write enriched records to stdout as CSV.

    Args:
        records:         Original market data records.
        figi_map:        instrument_id -> FIGI mapping.
        exchange_counts: Count of records per exchange.
        exchange_rank:   Rank of each exchange by frequency.
    """
    logger.info(f"Starting write_output for {len(records)} records")

    out_cols = [
        "record_id", "instrument_id", "id_type", "exchange_code",
        "figi", "price", "volume", "timestamp", "lookup_count", "exchange_rank",
    ]

    writer = csv.DictWriter(sys.stdout, fieldnames=out_cols, lineterminator="\n")
    writer.writeheader()

    for r in records:
        instrument_id = r.get("instrument_id") or ""
        exchange_code = r.get("exchange_code") or "UNKNOWN"
        writer.writerow({
            "record_id":     r.get("record_id") or "",
            "instrument_id": instrument_id,
            "id_type":       r.get("id_type") or "",
            "exchange_code": exchange_code,
            "figi":          figi_map.get(instrument_id, "UNKNOWN"),
            "price":         r.get("price") or "",
            "volume":        r.get("volume") or "",
            "timestamp":     r.get("timestamp") or "",
            "lookup_count":  exchange_counts.get(exchange_code, 0),
            "exchange_rank": exchange_rank.get(exchange_code, 0),
        })

    logger.info("Completed write_output")


def main(input_file: str) -> None:
    """Run the ingest pipeline.

    Args:
        input_file: Path string to the input CSV file.
    """
    logger.info(f"Starting ingest pipeline for {input_file}")

    input_path = Path(input_file)
    records = load_records(input_path)

    client = FigiClient()
    figi_map = resolve_figis(records, client)
    exchange_counts, exchange_rank = rank_exchanges(records)
    write_output(records, figi_map, exchange_counts, exchange_rank)

    logger.info("Ingest pipeline complete")


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(f"Usage: python {sys.argv[0]} <input_csv>", file=sys.stderr)
        sys.exit(1)
    main(sys.argv[1])

File: perl/ingest.pl
#!/usr/bin/perl

use strict;
use warnings;
use FindBin qw($Bin);
use lib $Bin;
use FigiClient;
use JSON qw(from_json to_json);
use File::Basename qw(basename);

=head1 NAME

ingest.pl -- Market data identifier ingestion pipeline

=head1 DESCRIPTION

Reads a CSV file of market data records, resolves each instrument
identifier to its OpenFIGI code using FigiClient, counts lookup
frequency by exchange code, and writes enriched records to STDOUT
as CSV.

Input CSV columns:
    record_id, instrument_id, id_type, exchange_code, price, volume, timestamp

Output CSV columns:
    record_id, instrument_id, id_type, exchange_code, figi, price, volume,
    timestamp, lookup_count

Usage:
    perl ingest.pl data/sample_input.csv > data/output.csv

=cut

my $input_file = $ARGV[0]
    or die "Usage: $0 <input_csv>\n";

open(my $fh, '<', $input_file)
    or die "Cannot open $input_file: $!\n";

my $header = <$fh>;
chomp $header;
my @cols = split(/,/, $header);

my @records;
while (my $line = <$fh>) {
    chomp $line;
    next unless length $line;
    my %rec;
    @rec{@cols} = split(/,/, $line, scalar @cols);
    push @records, \%rec;
}
close $fh;

die "ingest.pl: no records found in $input_file\n" unless @records;

# Build identifier batch for FigiClient
my @identifiers;
for my $rec (@records) {
    unless (defined $rec{instrument_id} && length $rec{instrument_id}) {
        warn "ingest.pl: skipping record $rec{record_id}: missing instrument_id\n";
        next;
    }
    push @identifiers, {
        idType => $rec{id_type} || 'TICKER',
        idValue => $rec{instrument_id},
        exchCode => $rec{exchange_code} || undef,
    };
}

# Resolve FIGIs
my $client = FigiClient->new(
    url    => 'https://api.openfigi.com/v2/mapping',
    apikey => $ENV{OPENFIGI_API_KEY} || 'DEMO_KEY',
);

my $figi_results = $client->do_request(\@identifiers);

# Build figi lookup by instrument_id
my %figi_map;
if (ref $figi_results eq 'ARRAY') {
    for my $i (0 .. $#identifiers) {
        my $id   = $identifiers[$i]{idValue};
        my $hits = $figi_results->[$i]{data};
        if (ref $hits eq 'ARRAY' && @$hits) {
            $figi_map{$id} = $hits->[0]{figi};
        }
        else {
            $figi_map{$id} = 'UNKNOWN';
        }
    }
}

# Count lookup frequency by exchange_code
my %exchange_counts;
for my $rec (@records) {
    my $exch = $rec{exchange_code} || 'UNKNOWN';
    $exchange_counts{$exch}++;
}

# Sort exchanges by count descending
my @sorted_exchanges = reverse sort { $exchange_counts{$a} <=> $exchange_counts{$b} }
                       keys %exchange_counts;

# Write output CSV
print join(',', qw(record_id instrument_id id_type exchange_code figi
                   price volume timestamp lookup_count exchange_rank)), "\n";

my %exchange_rank;
my $rank = 1;
for my $exch (@sorted_exchanges) {
    $exchange_rank{$exch} = $rank++;
}

for my $rec (@records) {
    my $id   = $rec{instrument_id} // '';
    my $exch = $rec{exchange_code} // 'UNKNOWN';
    my $figi = $figi_map{$id} // 'UNKNOWN';
    my $cnt  = $exchange_counts{$exch} // 0;
    my $rnk  = $exchange_rank{$exch} // 0;

    print join(',', (
        $rec{record_id}      // '',
        $id,
        $rec{id_type}        // '',
        $exch,
        $figi,
        $rec{price}          // '',
        $rec{volume}         // '',
        $rec{timestamp}      // '',
        $cnt,
        $rnk,
    )), "\n";
}
